### DISCUSSION TITLE
Get emoji or Win-V input

### DISCUSSION THREAD

**Discussion by MichaelRumpler** [AvaloniaUI/Avalonia#16550] [Q&A] [answered]
I have a custom control which handles all the key events.

		control.AddHandler(InputElement.KeyDownEvent, OnKeyDown, handledEventsToo: true);
		control.AddHandler(InputElement.KeyUpEvent, OnKeyUp, handledEventsToo: true);
		control.AddHandler(InputElement.TextInputEvent, OnTextInput, handledEventsToo: true);

In these events I get notified of all the keys which are pressed.

But when I press Windows-V to select some text from my clipboard history or Windows-. to select an emoji, then these do not send those events.

Is there some other event I'm not aware of where I can be notified about that?

When I press Win-V or Win-. in a `TextBox`, then the text is entered there. So somehow it must work.

**Comment by IanRawley** [comment]
At a guess, and I do mean guess, those key events are intercepted by another process to provide that functionality, preventing Avalonia from ever seeing them as Key Events.  If there's an IME involved, which by default there is if a Textbox is focused, it's likely receiving the event and converting it to a Text event.  I have similar conflicts between Avalonia apps and the AMD Adrenalin software intercepting certain key combos.

If you have IMEs involved KeyUp/Down events become very unreliable.  Especially so on Android, where the IME intercepts any key that could be text and turns it to a Text event in the absence of non-shift modifiers.

**Comment by MichaelRumpler** [comment]
I also handle the `TextInputEvent` and would've expected that to come in there.

**Comment by MichaelRumpler** [comment, marked as answer]
You were right, those two programs which pop up with Win-. and Win-V act like IMEs. I took a look at the [TextBox source](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBox.cs) and the [IME wiki page](https://github.com/AvaloniaUI/Avalonia/wiki/IME) and managed to add a simple IME client to my code.

Here is the essence of it:

	internal partial class KeyboardHandler : IDisposable
	{
		private Control Control;
		private readonly IMClient _imClient;


		public KeyboardHandler(Control control, Func<Rect> getCursorRect)
		{
			Control = control;
			_imClient = new(Control, getCursorRect);

			control.AddHandler(InputElement.KeyDownEvent, OnKeyDown, handledEventsToo: true);
			control.AddHandler(InputElement.TextInputEvent, OnTextInput, handledEventsToo: true);
			control.AddHandler(InputElement.TextInputMethodClientRequestedEvent, OnIMERequested, handledEventsToo: true);
		}

		private void OnIMERequested(object? sender, TextInputMethodClientRequestedEventArgs e)
		{
			e.Client = _imClient;
		}

		class IMClient : TextInputMethodClient
		{
			private Control Control;
			private readonly Func<Rect> GetCursorRect;

			public IMClient(Control control, Func<Rect> getCursorRect)
			{
				Control = control;
				GetCursorRect = getCursorRect;
			}

			public override Visual TextViewVisual => Control;
			public override Rect CursorRectangle => GetCursorRect();

			public override bool SupportsPreedit => false;
			public override bool SupportsSurroundingText => false;
			public override string SurroundingText => string.Empty;		// may return the text of the current line
			public override TextSelection Selection { get => new(); set { } }
		}

		private async void OnKeyDown(object? sender, KeyEventArgs e) ...

		private void OnTextInput(object? sender, TextInputEventArgs e) ...
	}

So in the `TextInputMethodClientRequestedEvent` I need to return a subclass of TextInputMethodClient. In my case it didn't need to do very much. Just return the Control in `TextViewVisual` and the cursor position in `CursorRectangle`. Everything else returns default values.

The [TextBoxTextInputMethodClient](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBoxTextInputMethodClient.cs) also handles selection and caret position, but in my case I didn't need that.

Now Win-. does raise the `TextInputEvent` with the selected emoji. Win-V changes the clipboard content and raises `KeyDownEvent` with Ctrl-V. So I just had to paste the clipboard in the `OnKeyDown` method.

Unfortunately I cannot test it with some Asian keyboard, but that should at least be better now too.
